### PR TITLE
Expand the Prior Literature after additional literature review

### PR DESCRIPTION
**Pull request by willy-b**
Paper draft link: https://raw.githubusercontent.com/willy-b/RASP-for-ReCOGS/e41b1af351226109987994a4aaa4ade570c14cb5/rasp-for-recogs_pos-wbruns-2024-draft.pdf

Need to move a bunch of stuff to appendices and cut length of the paper and the abstract again soon, but want to checkpoint here.

Add references to:
- Brenden M Lake and Marco Baroni. 2023. Human-like systematic generalization through a meta-learning neural network. Nature, 623(7985):115–121
They use a "meta-learning for compositionality" framework with a 3-layer Encoder-Decoder Transformer and get 100% on COGS, except the structural generalization splits we focus on in RASP-for-ReCOGS where their model still gets 0% accuracy (they excel at other benchmarks and COGS is not the sole focus of their paper but it is of interest to mention).

- Jackson Petty, Sjoerd van Steenkiste, Ishita Dasgupta, Fei Sha, Dan Garrette, and Tal Linzen. 2024. The impact of depth on compositional generalization in transformer language models.
"We find that even deep models at the largest model size systematically fail to generalize structurally (Figure 5); the benefit of depth is largely limited to the easier lexical generalization cases. This supports the hypothesis that the saturated effect of depth is due to the existence of easier subsets of the datasets, and shows that increasing depth alone does not substantially improve the models’ ability to learn the correct inductive bias for these structural tasks."

- Bingzhi Li, Lucia Donatelli, Alexander Koller, Tal Linzen, Yuekun Yao, and Najoung Kim. 2023. Slog: A structural generalization benchmark for semantic parsing.
Extends COGS benchmark with a large number of additional challenging structural generalization cases. Adds the split we independently predicted and verified as difficult in this paper based RASP analysis (before seeing this paper), "v_dat_p2_pp_moved_to_recipient", and also mentions very briefly with no analysis that they also see nearest noun substitution in certain obj_pp_to_subj_pp split pp generalization errors (which we predict and analyze here and show we can predict the details of 96% of the time across n=10 separately trained from scratch Transformers running on the task). 

- O. Jespersen. 1954. A Modern English Grammar on Historical Principles: Volume 2, Syntax (first volume). Otto Jespersen. Bradford and Dickens.
Note that such nearest noun "attraction" errors we report here where attractor/distractor prepositional phrase nouns replace the actual agent/subject in the ReCOGS logical form generated by Wu et al 2023 baseline Transformers are NOT due to their presence in pre-training or training data, as the ReCOGS/COGS training data is synthetic and syntactically perfect and for this benchmark the Transformer is trained from scratch, so it a genuine new error made by the neural network itself (and we predict a mechanism using RASP). But in general, _humans do also exhibit these "attraction" errors, e.g. in subject-verb agreement per Jespersen "Very frequently in speech, and not infrequently in literature, the number of the verb is determined by that part of the subject which is nearest to the verb, even if a stricter sense of grammar would make the verb agree with the main part of the subject. This kind of attraction naturally occurs the more easily, the greater the distance is between the nominative and the verb"_, so pre-trained models trained on human-generated text may have the additional problem of learning those errors from the training data itself.
Language model causality issues (in causal decoders) like trying to output a long addition digit by digit starting with the most significant digit in a long addition before computing the sums of the less significant digits that come later, and failing if there is a carry at any point, are not limited to math, nor limited to language models, for just one example from English grammar concerning human language processing, _Jespersen explains "Concord of the verb" errors made by humans especially in speech when the verb is on the left due to needing to agree with a noun not explicitly selected yet: "The general rule, which needs no exemplification, is for the verb to be in the singular with a singular subject, and in the plural with a plural subject. Occasionally, however, the verb will be put in the [singular], even if the subject is plural; this will especially happen when the verb precedes the subject, because the speaker has not yet made up his mind, when pronouncing the verb, what words are to follow."_

- Arnold Zwicky. 2008. Agreement with nearest. https://languagelog.ldc.upenn.edu/nll/?p=839
Interesting reading from an expert linguist cited by Linzen and Goldberg, led me to Jespersen by way of him quoting part of the Merriam-Webster Dictionary of English Usage that quoted Jespersen (not either of the quotes used above or even the same specific section as I ended up using but that is why I was reading Jespersen, and subject-verb agreement is also discussed in this article as is a nearest attraction error).

- Tal Linzen, Emmanuel Dupoux, and Yoav Goldberg. 2016. Assessing the ability of lstms to learn syntax-
sensitive dependencies. Transactions of the Association for Computational Linguistics, 4:521–535.
Relevant for discussion of hierarchical/tree-model relevant aspects of syntax and "attraction" errors where long-range dependencies are distracted by closer attractor nouns in the context of subject-verb agreement instead of prepositional phrase modification.

Another good pair of papers from same authors on Transformers vs compositional syntax, with special attention to subject-verb agreement nearest noun attraction errors:
- Marten van Schijndel, Aaron Mueller, and Tal Linzen. 2019. Quantity doesn’t buy quality syntax with neural language models.
and
- Yoav Goldberg. 2019. Assessing BERT’s syntactic abilities. (In depth and useful)

Lastly,
- John Hewitt and Christopher D. Manning. 2019. A structural probe for finding syntax in word representations. In Proceedings of the 2019 Conference of the North American Chapter of the Association for
Computational Linguistics: Human Language Technologies, Volume 1 (Long and Short Papers), pages
4129–4138, Minneapolis, Minnesota. Association for Computational Linguistics.
They argue there is evidence entire compositional parse trees for complex natural sentences are embedded in the vector space representation of deeper layers of BERT Transformers.

Credit to @i-am-ying-li who has not been involved in this paper so far but who reminded me of Hewitt and Manning 2019 ( https://aclanthology.org/N19-1419/ ) and Andreas 2019 ( https://arxiv.org/abs/1902.07181 ) which were discussed in Stanford XCS224N in relation to compositional generalization. 

Hewitt and Manning 2019 was added to the literature review for this paper, Andreas 2019 may be mentioned somewhere as it is excellent reading on compositional generalization though it is focused on the embedding/vector space representation which is not a focus of this paper.

### CODE REVIEW

**Comment by willy-b** [comment]
(Description has been updated to include this information but noting the change here so it goes into my log)

Add references to:
- Brenden M Lake and Marco Baroni. 2023. Human-like systematic generalization through a meta-learning neural network. Nature, 623(7985):115–121
They use a "meta-learning for compositionality" framework with a 3-layer Encoder-Decoder Transformer and get 100% on COGS except the structural generalization splits we focus on in RASP-for-ReCOGS where their model still gets 0% accuracy (they excel at other benchmarks and COGS is not the sole focus of their paper but it is of interest to mention).

- Jackson Petty, Sjoerd van Steenkiste, Ishita Dasgupta, Fei Sha, Dan Garrette, and Tal Linzen. 2024. The impact of depth on compositional generalization in transformer language models.
"We find that even deep models at the largest model size systematically fail to generalize structurally (Figure 5); the benefit of depth is largely limited to the easier lexical generalization cases. This supports the hypothesis that the saturated effect of depth is due to the existence of easier subsets of the datasets, and shows that increasing depth alone does not substantially improve the models’ ability to learn the correct inductive bias for these structural tasks."

- Bingzhi Li, Lucia Donatelli, Alexander Koller, Tal Linzen, Yuekun Yao, and Najoung Kim. 2023. Slog: A structural generalization benchmark for semantic parsing.
Extends COGS benchmark with a large number of additional challenging structural generalization cases. Adds the split we independently predicted and verified as difficult in this paper based RASP analysis (before seeing this paper), "v_dat_p2_pp_moved_to_recipient", and also mentions very briefly with no analysis that they also see nearest noun substitution in certain obj_pp_to_subj_pp split pp generalization errors (which we predict and analyze here and show we can predict the details of 96% of the time across n=10 separately trained from scratch baseline Transformers running on the task). 

- O. Jespersen. 1954. A Modern English Grammar on Historical Principles: Volume 2, Syntax (first volume). Otto Jespersen. Bradford and Dickens.
Note that such nearest noun "attraction" errors we report here where attractor/distractor prepositional phrase nouns replace the actual agent/subject in the ReCOGS logical form generated by Wu et al 2023 baseline Transformers are NOT due to their presence in pre-training or training data, as the ReCOGS/COGS training data is synthetic and syntactically perfect and for this benchmark the Transformer is trained from scratch, so it a genuine new error made by the neural network itself (and we predict a mechanism using RASP). But in general, _humans do also exhibit these "attraction" errors, e.g. in subject-verb agreement per Jespersen "Very frequently in speech, and not infrequently in literature, the number of the verb is determined by that part of the subject which is nearest to the verb, even if a stricter sense of grammar would make the verb agree with the main part of the subject. This kind of attraction naturally occurs the more easily, the greater the distance is between the nominative and the verb"_, so pre-trained models trained on human-generated text may have the additional problem of learning those errors from the training data itself.
Language model causality issues (in causal decoders) like trying to output a long addition digit by digit starting with the most significant digit in a long addition before computing the sums of the less significant digits that come later, and failing if there is a carry at any point, are not limited to math, nor limited to language models, for just one example from English grammar concerning human language processing, _Jespersen explains "Concord of the verb" errors made by humans especially in speech when the verb is on the left due to needing to agree with a noun not explicitly selected yet: "The general rule, which needs no exemplification, is for the verb to be in the singular with a singular subject, and in the plural with a plural subject. Occasionally, however, the verb will be put in the [singular], even if the subject is plural; this will especially happen when the verb precedes the subject, because the speaker has not yet made up his mind, when pronouncing the verb, what words are to follow."_

- Arnold Zwicky. 2008. Agreement with nearest. https://languagelog.ldc.upenn.edu/nll/?p=839
Interesting reading from an expert linguist cited by Linzen and Goldberg, led me to Jespersen by way of him quoting part of the Merriam-Webster Dictionary of English Usage that quoted Jespersen (not either of the quotes used above or even the same specific section as I ended up using but that is why I was reading Jespersen, and subject-verb agreement is also discussed in this article as are some nearest attraction errors).

- Tal Linzen, Emmanuel Dupoux, and Yoav Goldberg. 2016. Assessing the ability of lstms to learn syntax-
sensitive dependencies. Transactions of the Association for Computational Linguistics, 4:521–535.
Relevant for discussion of hierarchical/tree-model relevant aspects of syntax and "attraction" errors where long-range dependencies are distracted by closer attractor nouns in the context of subject-verb agreement instead of prepositional phrase modification.

Another good pair of papers from same authors on Transformers vs compositional syntax, with special attention to subject-verb agreement nearest noun attraction errors:
- Marten van Schijndel, Aaron Mueller, and Tal Linzen. 2019. Quantity doesn’t buy quality syntax with neural language models.
and
- Yoav Goldberg. 2019. Assessing BERT’s syntactic abilities. (In depth and useful)

Lastly, 
- John Hewitt and Christopher D. Manning. 2019. A structural probe for finding syntax in word representations. In Proceedings of the 2019 Conference of the North American Chapter of the Association for
Computational Linguistics: Human Language Technologies, Volume 1 (Long and Short Papers), pages
4129–4138, Minneapolis, Minnesota. Association for Computational Linguistics.
They argue there is evidence entire compositional parse trees for complex natural sentences are embedded in the vector space representation of deeper layers of BERT Transformers.

**Comment by willy-b** [comment]
Note the two articles I quoted short excerpts from above are both liberally licensed for such reuse and open access (and I cited them fully):
- Jackson Petty, Sjoerd van Steenkiste, Ishita Dasgupta, Fei Sha, Dan Garrette, and Tal Linzen. 2024. The impact of depth on compositional generalization in transformer language models.
 is from https://arxiv.org/abs/2310.19956 and is licensed Creative Commons BY 4.0 https://creativecommons.org/licenses/by/4.0/
- Brenden M Lake and Marco Baroni. 2023. Human-like systematic generalization through a meta-learning neural network. Nature, 623(7985):115–121
 https://www.nature.com/articles/s41586-023-06668-3#rightslink is also Creative Commons BY 4.0: http://creativecommons.org/licenses/by/4.0/

I do use e.g. copyright.com RightsLink to request or buy reuse rights where such it is not licensed so liberally.
